src/operator/csv.rs

```diff
@@ -40,52 +44,74 @@ where
     pub fn from_csv_reader(reader: CsvReader<R>) -> Self {
         Self {
             reader,
+            time: 0,
             _t: PhantomData,
         }
     }
 }
 
-impl<R, C, T> Operator for CsvSource<R, C, T>
+impl<R, T, W, C> Operator for CsvSource<R, T, W, C>
 where
     C: Data,
     R: 'static,
     T: 'static,
+    W: 'static,
 {
     fn name(&self) -> Cow<'static, str> {
         Cow::from("CsvSource")
     }
-    fn clock_start(&mut self, _scope: Scope) {}
+    fn clock_start(&mut self, _scope: Scope) {
+        self.time = 0;
+    }
     fn clock_end(&mut self, _scope: Scope) {}
+    fn fixedpoint(&self) -> bool {
+        self.time >= 2
+    }
 }
 
-impl<R, C, T> SourceOperator<C> for CsvSource<R, C, T>
+impl<R, T, W, C> SourceOperator<C> for CsvSource<R, T, W, C>
 where
     T: for<'de> Deserialize<'de> + 'static,
+    W: ZRingValue + 'static,
     R: Read + 'static,
-    C: Data + FromIterator<T>,
+    C: Data + ZSet<Key = T, R = W>,
 {
     fn eval(&mut self) -> C {
-        self.reader.deserialize().map(CsvResult::unwrap).collect()
+        if self.time == 0 {
+            let input_iter = self.reader.deserialize().map(CsvResult::unwrap).into_iter();
+
+            let mut data = Vec::with_capacity(input_iter.size_hint().0);
+
+            for x in input_iter {
+                data.push(((x, ()), W::one()));
+            }
+
+            self.time += 1;
+            C::from_tuples((), data)
```

Comment:
isn't there a way to do this without a vector, just using an iterator?

Reply:
The current Batcher API will actually reuse the heap allocation behind this vector, so this is deliberate, at least in the DD design.